File: YASAudio/Graph/YASAudioUnitParameter.m
//
//  YASAudioUnitParameter.m
//  Copyright (c) 2015 Yuki Yasoshima.
//

#import "YASAudioUnitParameter.h"
#import "YASMacros.h"
#import "NSString+YASAudio.h"

@implementation YASAudioUnitParameter

- (instancetype)initWithAudioUnitParameterInfo:(const AudioUnitParameterInfo *)info
                                   parameterID:(const AudioUnitParameterID)parameterID
                                         scope:(const AudioUnitScope)scope
{
    self = [super init];
    if (self) {
        _parameterID = parameterID;
        _scope = scope;

        if (info->unitName != NULL) {
            _unitName = YASRetain((__bridge NSString *)(info->unitName));
        }

        if (info->cfNameString != NULL) {
            _name = YASRetain((__bridge NSString *)(info->cfNameString));
        }

        _hasClump = info->flags & kAudioUnitParameterFlag_HasClump;
        _clumpID = info->clumpID;
        _unit = info->unit;
        _minValue = info->minValue;
        _maxValue = info->maxValue;
        _defaultValue = info->defaultValue;
        _values = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)dealloc
{
    YASRelease(_unitName);
    YASRelease(_name);
    YASRelease(_values);

    _unitName = nil;
    _name = nil;
    _values = nil;

    YASSuperDealloc;
}

- (NSString *)description
{
    NSMutableString *result = [NSMutableString stringWithFormat:@"<%@: %p>\n{\n", self.class, self];
    NSMutableArray *lines = [NSMutableArray array];
    [lines addObject:[NSString stringWithFormat:@"UnitName = %@", _unitName]];
    [lines addObject:[NSString stringWithFormat:@"Name = %@", _name]];
    [lines addObject:[NSString stringWithFormat:@"ParameterID = %@", @(_parameterID)]];
    [lines addObject:[NSString stringWithFormat:@"Scope = %@", [NSString yas_stringWithAudioUnitScope:_scope]]];
    [lines addObject:[NSString stringWithFormat:@"HasClamp = %@", @(_hasClump)]];
    [lines addObject:[NSString stringWithFormat:@"ClumpID = %@", @(_clumpID)]];
    [lines addObject:[NSString stringWithFormat:@"Unit = %@", [NSString yas_stringWithAudioUnitParameterUnit:_unit]]];
    [lines addObject:[NSString stringWithFormat:@"MinValue = %@", @(_minValue)]];
    [lines addObject:[NSString stringWithFormat:@"MaxValue = %@", @(_maxValue)]];
    [lines addObject:[NSString stringWithFormat:@"DefaultValue = %@", @(_defaultValue)]];
    [result appendString:[[lines componentsJoinedByString:@"\n"] stringByAppendingLinePrefix:@"    "]];
    [result appendFormat:@"\n}"];
    return result;
}

- (Float32)valueForElement:(const AudioUnitElement)element
{
    NSNumber *valueNumber = _values[@(element)];
    if (valueNumber) {
        return valueNumber.floatValue;
    } else {
        return _defaultValue;
    }
}

- (void)setValue:(Float32)value forElement:(const AudioUnitElement)element
{
    _values[@(element)] = @(value);
}

@end

File: YASAudio/Graph/yas_audio_unit_parameter.cpp
//
//  yas_audio_unit_parameter.cpp
//  Copyright (c) 2015 Yuki Yasoshima.
//

#include "yas_audio_unit_parameter.h"
#include "yas_cf_utils.h"
#include "YASMacros.h"

using namespace yas;

class audio_unit_parameter::impl
{
   public:
    AudioUnitParameterID parameter_id;
    AudioUnitScope scope;
    bool has_clump;
    UInt32 clump_id;
    AudioUnitParameterUnit unit;
    AudioUnitParameterValue min_value;
    AudioUnitParameterValue max_value;
    AudioUnitParameterValue default_value;
    std::map<AudioUnitElement, AudioUnitParameterValue> values;
    std::string unit_name;
    std::string name;

    impl(const AudioUnitParameterInfo &info, const AudioUnitParameterID parameter_id, const AudioUnitScope scope)
        : parameter_id(parameter_id),
          scope(scope),
          has_clump(info.flags & kAudioUnitParameterFlag_HasClump),
          clump_id(info.clumpID),
          unit(info.unit),
          min_value(info.minValue),
          max_value(info.maxValue),
          default_value(info.defaultValue),
          values(),
          unit_name(yas::to_string(info.unitName)),
          name(yas::to_string(info.cfNameString))
    {
    }

    ~impl()
    {
    }

    impl(impl &&impl)
    {
        parameter_id = std::move(impl.parameter_id);
        scope = std::move(impl.scope);
        has_clump = std::move(impl.has_clump);
        clump_id = std::move(impl.clump_id);
        unit = std::move(impl.unit);
        min_value = std::move(impl.min_value);
        max_value = std::move(impl.max_value);
        default_value = std::move(impl.default_value);
        values = std::move(impl.values);
        unit_name = std::move(impl.unit_name);
        name = std::move(impl.name);
    }

    impl &operator=(impl &&impl)
    {
        return *this;
    }
};

audio_unit_parameter::audio_unit_parameter(const AudioUnitParameterInfo &info, const AudioUnitParameterID parameter_id,
                                           const AudioUnitScope scope)
    : _impl(std::make_unique<impl>(info, parameter_id, scope))
{
}

audio_unit_parameter::~audio_unit_parameter() = default;

audio_unit_parameter::audio_unit_parameter(audio_unit_parameter &&parameter)
{
    _impl = std::move(parameter._impl);
}

audio_unit_parameter &audio_unit_parameter::operator=(audio_unit_parameter &&parameter)
{
    if (this == &parameter) {
        return *this;
    }
    _impl = std::move(parameter._impl);
    return *this;
}

#pragma mark - accessor

AudioUnitParameterID audio_unit_parameter::parameter_id() const
{
    return _impl->parameter_id;
}

AudioUnitScope audio_unit_parameter::scope() const
{
    return _impl->scope;
}

const std::string &audio_unit_parameter::unit_name() const
{
    return _impl->unit_name;
}

bool audio_unit_parameter::has_clump() const
{
    return _impl->has_clump;
}

UInt32 audio_unit_parameter::clump_id() const
{
    return _impl->clump_id;
}

const std::string &audio_unit_parameter::name() const
{
    return _impl->name;
}

AudioUnitParameterUnit audio_unit_parameter::unit() const
{
    return _impl->unit;
}

AudioUnitParameterValue audio_unit_parameter::min_value() const
{
    return _impl->min_value;
}

AudioUnitParameterValue audio_unit_parameter::max_value() const
{
    return _impl->max_value;
}

AudioUnitParameterValue audio_unit_parameter::default_value() const
{
    return _impl->default_value;
}

Float32 audio_unit_parameter::value(const AudioUnitElement element) const
{
    return _impl->values.at(element);
}

void audio_unit_parameter::set_value(const AudioUnitParameterValue value, const AudioUnitElement element)
{
    _impl->values[element] = value;
}

const std::map<AudioUnitElement, AudioUnitParameterValue> &audio_unit_parameter::values() const
{
    return _impl->values;
}
